nexus/bin/qmca

```diff
@@ -1842,16 +1845,55 @@ class DatAnalyzer(QBase):
             v = data.LocalEnergy_sq - data.LocalEnergy**2
             data.Variance  = v
             stats.Variance = self.stat_value(v[nbe:])
-        #end if            
-        if 'LocalEnergy' in data and 'ElecElec' in data and 'MPC' in data and 'KEcorr' in data:
-            v = data.LocalEnergy - data.ElecElec + data.MPC + data.KEcorr
-            data.CorrectedEnergy  = v
-            stats.CorrectedEnergy = self.stat_value(v[nbe:])
-        elif 'LocalEnergy' in data and 'ElecElec' in data and 'MPC' in data:
-            v = data.LocalEnergy - data.ElecElec + data.MPC
-            data.CorrectedEnergy  = v
-            stats.CorrectedEnergy = self.stat_value(v[nbe:])
         #end if
+
+        # find finite size corrected energy
+        if 'LocalEnergy' in data and ('MPC' in data or 'KEcorr' in data):
+            ce = None
+            if 'Kinetic' in data:
+                non_ee_energies = self.non_ee_potentials
+                nee = data.Kinetic.copy()
+                for pot in self.non_ee_potentials:
+                    if pot in data:
+                        nee += data[pot]
+                    #end if
+                #end for
+                # allow the correction only if the local energy sum is correct
+                le_recovered = False
+                le_ref = data.LocalEnergy.sum()
+                for ee in ('ElecElec','MPC'):
+                    if ee in data:
+                        le_sum = (nee + data[ee]).sum()
+                        le_recovered |= abs((le_sum-le_ref)/le_ref)<1e-8
+                    #end if
+                #end if
+                if le_recovered:
+                    ce = nee
+                    if 'MPC' in data:
+                        ce += data.MPC
+                    elif 'ElecElec' in data:
+                        ce += data.ElecElec
+                    #end if
+                    if 'KEcorr' in data:
+                        ce += data.KEcorr
+                    #end if
+                #end if
+            #end if
+            if ce is None:
```

Comment:
if condition based on `ce` as `none` or not good. Taking me 10 minutes to understand the logic makes me feel it is suboptimal here. Why not just define a flag MPC_is_physical if nee+MPC=le.

if MPC_is_physical 
  ce = le
else if MPC in data and ElecElec in data
  ce = le +MPC - ElecElec
end if

if KEcorr in data
  ce += KEcorr
endif

Reply:
I also don't get why nee needs to computed based on Kinetic.
if ElecElec in data
  nee = LE - ElecElec
else
  nee = Le
endif

if MPC in data and (nee + MPC == LE)
  MPC_is_physical =true
endif

Reply:
The approach taken by the "if 'Kinetic' in data" block requires no special knowledge of the terms QMCPACK used internally to form the local energy.  It checks if it can recover the local energy by using either ElecElec or MPC as the e-e energy just to confirm that all the other terms included via self.non_ee_potentials comprehensively represents the non-e-e potential terms.  If it can, the corrected energy is always kinentic + non-e-e + MPC (if present, and +ElecElec otherwise).

Reply:
The 'if ce is None' is a fallback in case this approach does not work, e.g. a potential term is missing in the list.  The fallback represents the approach taken in the code prior to this PR, which is likely only correct if ElecElec is physical and MPC is not.

Reply:
See updates now with in-code comments.